### PR TITLE
Add device memory tracker as backend utilties

### PR DESCRIPTION
**Pull request by GuanLuo**

### CODE REVIEW

File: include/triton/backend/device_memory_tracker.h [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+/// TRITONBACKEND_BackendAttributeSetEnableMemoryTracker to true, which provides
+/// additional information to Triton for making decision on model scaling and
+/// deployment.
+///
+/// Caveat: The memory tracker is implemented with CUPTI library which currently
+/// only supports single client/subscriber. This is an known limitation and as a
+/// result, the memory tracker can cause unexpected application failure if other
+/// component of the Triton process also uses CUPTI with a different
+/// configuration, for example, the framework used by the backend may have
+/// implemented similar profiler with CUPTI. Therefore, before enabling this
```
Comment (rmccorm4) [inline include/triton/backend/device_memory_tracker.h:53]: Is it possible for the DeviceMemoryTracker internally to check if there are other CUPTI clients active via some CUPTI APIs?

Comment (GuanLuo) [reply include/triton/backend/device_memory_tracker.h:53]: See https://github.com/triton-inference-server/backend/pull/80/files#diff-f8ad565984e1abf1676a6d0b85f921cb0a29f78f6532debe29d2ca8d110dd51dR108-R127

File: include/triton/backend/device_memory_tracker.h [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+extern "C" {
+
+typedef struct TRITONBACKEND_CuptiTracker_t {
+  // C struct require extra implementation for dynamic array, for simplicity,
+  // the following assumptions are made to pre-allocate the array with max
+  // possible length:
+  //  - system / pinned memory allocation should only be on deviceId 0
+  //  - CUDA allocation will only be on visible CUDA devices
+  int64_t* system_byte_size_;
```
Comment (nnshah1) [inline include/triton/backend/device_memory_tracker.h:99]: naming here is a little confusing at first glance as it indicates a single "size", could we use something like:
```
system_memory_usage
cuda_memory_usage
```

File: include/triton/backend/device_memory_tracker.h [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+  static bool Init();
+
+  static int CudaDeviceCount();
+
+  // The memory usage will be tracked and modified until it's untracked, 'usage'
+  // must be valid and not to be modified externally until untrack is called.
+  // Currently can distinguish activity by correlation id which is thread
+  // specific, which implies that there will be mssing records if tracking
+  // region switching threads to handle other activities.
+  static void TrackThreadMemoryUsage(MemoryUsage* usage);
```
Comment (nnshah1) [inline include/triton/backend/device_memory_tracker.h:225]: Should this be the C struct? and not the C++ ? TRITONBACKEND_CuptiTracker_t

Also maybe we could rename to TRITONBACKEND_MemoryUsageTracker_t - to avoid explicitly referencing Cupti

File: src/device_memory_tracker.cc [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+  case CUPTI_SUCCESS : {
+    CUPTI_CALL(
+        cuptiActivityRegisterCallbacks(bufferRequested, bufferCompleted));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_MEMORY2));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+    break;
+  }
+  case CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED: {
+    std::cerr << "CUPTI has been initialized elsewhere, assuming the implementation is the same" << std::endl;
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:132]: Is there a way to check that those activities are enabled? If not would it be better to fail here?

File: include/triton/backend/device_memory_tracker.h (hunk 3 above)
Comment (GuanLuo) [reply include/triton/backend/device_memory_tracker.h:225]: The C struct is what actually being passed to the CUPTI callback (user really shouldn't interact directly), `MemoryUsage` and `DeviceMemoryTracker` is the C++ class / struct that the user should interacts with.

I could be wrong as I am not an expert on the subject, but I don't think we can get rid of the C struct for ABI compatibility reason, the actual CUPTI callback and the structure used in the callback is C to ensure the (de facto) ABI is the same. Note that the ABI concern comes into the picture as CUPTI may have only one callback register (i.e. from backend A) and used by all backends which can be compiled with different C++ compiler.

File: src/device_memory_tracker.cc [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    throw std::runtime_error("DeviceMeoryTracker::Init() must be called before using any DeviceMeoryTracker features.");
+  }
+}
+
+void
+DeviceMemoryTracker::UntrackThreadMemoryUsage(MemoryUsage* usage)
+{
+  if (tracker_) {
+    uint64_t id;
+    CUPTI_CALL(cuptiActivityPopExternalCorrelationId(
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:182]: qq: would we want to flush first? and then pop?

File: src/device_memory_tracker.cc [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    case CUPTI_ACTIVITY_KIND_MEMORY2: {
+      CUpti_ActivityMemory3* memory_record = (CUpti_ActivityMemory3*)record;
+      TRITONBACKEND_CuptiTracker* usage = nullptr;
+      {
+        std::lock_guard<std::mutex> lk(mtx_);
+        auto it = activity_to_memory_usage_.find(memory_record->correlationId);
+        if (it != activity_to_memory_usage_.end()) {
+          usage = reinterpret_cast<TRITONBACKEND_CuptiTracker*>(it->second);
+          activity_to_memory_usage_.erase(it);
+        }
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:204]: instead of the ```if (usage !=nullptr``` - maybe we can return here if the it == end() - just to reduce indentation level below

File: src/device_memory_tracker.cc [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+      // Ignore memory record that is not associated with a TRITONBACKEND_CuptiTracker
+      // object
+      if (usage != nullptr) {
+        const bool is_allocation =
+            (memory_record->memoryOperationType ==
+             CUPTI_ACTIVITY_MEMORY_OPERATION_TYPE_ALLOCATION);
+        const bool is_release =
+            (memory_record->memoryOperationType ==
+             CUPTI_ACTIVITY_MEMORY_OPERATION_TYPE_RELEASE);
+        if (is_allocation || is_release) {
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:215]: similar maybe an early return here would be clearer in case its not allocation or release -

File: src/device_memory_tracker.cc [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+          switch (memory_record->memoryKind) {
+            case CUPTI_ACTIVITY_MEMORY_KIND_DEVICE: {
+              if (memory_record->deviceId >= usage->cuda_array_len_) {
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:218]: since the logic here is the same, - could refactor to an inline function taking the array and arraylen and memory_record.

Another option could be to set local variables (i.e. auto memory_usage_array = usage->cuda_byte_size, memory_usage_array_len = usage->cuda_array_len.

File: include/triton/backend/device_memory_tracker.h [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+      std::plus<int64_t>());
+      std::transform(rhs.cuda_byte_size_.begin(), rhs.cuda_byte_size_.end(), cuda_byte_size_.begin(), cuda_byte_size_.begin(),
+      std::plus<int64_t>());
+      return *this;
+    }
+
+    TRITONSERVER_Error* SerializeToBufferAttributes(
+      TRITONSERVER_BufferAttributes** usage, int32_t* usage_size)
+    {
+      int32_t usage_idx = 0;
```
Comment (nnshah1) [inline include/triton/backend/device_memory_tracker.h:156]: should this fail if the cupti record is invalid?

File: include/triton/backend/device_memory_tracker.h [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+  int64_t* system_byte_size_;
+  int64_t* pinned_byte_size_;
+  int64_t* cuda_byte_size_;
+  uint32_t system_array_len_;
+  uint32_t pinned_array_len_;
+  uint32_t cuda_array_len_;
+
+  // only set to false if somehow the CUPTI activity occurs on index out of
+  // range. In that case, user should invalidate the whole tracker.
+  bool good_record_;
```
Comment (nnshah1) [inline include/triton/backend/device_memory_tracker.h:108]: instead of `good_record_` maybe `valid` would be clearer?

File: include/triton/backend/device_memory_tracker.h (hunk 3 above)
Comment (nnshah1) [reply include/triton/backend/device_memory_tracker.h:225]: > The C struct is what actually being passed to the CUPTI callback (user really shouldn't interact directly), `MemoryUsage` and `DeviceMemoryTracker` is the C++ class / struct that the user should interacts with.
> 
> I could be wrong as I am not an expert on the subject, but I don't think we can get rid of the C struct for ABI compatibility reason, the actual CUPTI callback and the structure used in the callback is C to ensure the (de facto) ABI is the same. Note that the ABI concern comes into the picture as CUPTI may have only one callback register (i.e. from backend A) and used by all backends which can be compiled with different C++ compiler.

The MemoryUsage class though is transparent to CUPTI (i.e. it is just the external correlation ID). It doesn't dereference it. As the callback is in our C++ code - then I think we would be safe to get rid of it. However, I thought the backend API is C - so thought the the backend utilities would be C as well. -

Comment (nnshah1) [reply include/triton/backend/device_memory_tracker.h:225]: I see that that each backend statically links to the backend utils library - but this is the first case where we have an extern global - I'm not sure in this case exactly how things will resolve. I don't believe however that the C struct itself will handle ABI between backends. Will discuss offline.

Comment (nnshah1) [reply include/triton/backend/device_memory_tracker.h:225]: resolved in offline discussion - C struct used here to ensure any memory tracker singleton can be used independent of compiler

File: src/device_memory_tracker.cc [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+  if (tracker_) {
+    return tracker_->device_cnt_;
+  }
+  return 0;
+}
+
+bool
+DeviceMemoryTracker::Init()
+{
+  if (tracker_ == nullptr) {
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:153]: During local testing - I found this to not be thread safe in the case of calling Init() from a ModelInitialize(). So that if you have a backend responsible for multiple models and it calls Init() - there's a chance that both will call DeviceMemoryTracker(). Only one will succeed in subscribing - but the other one will be destroyed and can lead to confusion. Suggest a mutex

File: src/device_memory_tracker.cc [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+              break;
+          }
+        }
+      }
+      break;
+    }
+    case CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION: {
+      CUpti_ActivityExternalCorrelation* corr =
+          (CUpti_ActivityExternalCorrelation*)record;
+      std::lock_guard<std::mutex> lk(mtx_);
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:272]: can we add a check for externalcorrelationkind ==CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN

File: src/device_memory_tracker.cc (hunk 11 above)
Comment (GuanLuo) [reply src/device_memory_tracker.cc:153]: `DeviceMemoryTracker::Init()` should be called on `TRITONBACKEND_Initialize()`

Comment (nnshah1) [reply src/device_memory_tracker.cc:153]: thanks!

File: src/device_memory_tracker.cc [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+  }
+}
+
+int
+DeviceMemoryTracker::CudaDeviceCount()
+{
+  if (tracker_) {
+    return tracker_->device_cnt_;
+  }
+  return 0;
```
Comment (nnshah1) [inline src/device_memory_tracker.cc:147]: can we add `throw std::runtime_error("DeviceMemoryTracker::Init() must be called before using any DeviceMeoryTracker features.");`

**Comment by nnshah1** [comment]
Overall looks good! A few thoughts after playing around with it some:

- Unsubscribe doesn't unregister the callbacks (I think this is just a limitation of cupti)
- If a client subscribes before us - then tracking won't work - and unfortunately we can't tell.
- If a client pushes a different external correlation id - we would seg fault
- If we subscribe after someone else - then tracking won't work and unfortunately we can't tell.

I think the only way to protect ourselves would be to move the initialization and tracking to the server / core (though have backends call in to manage lifetimes / etc.) - but this is a bigger redesign.

File: src/device_memory_tracker.cc (hunk 13 above)
Comment (GuanLuo) [reply src/device_memory_tracker.cc:147]: No, this is not throw intentionally, `MemoryUsage` will access this function and we will have to do a lot more handling if the backend disable the tracking.

Comment (nnshah1) [reply src/device_memory_tracker.cc:147]: but the ordering is important - i.e. if you have a MemoryUsage object that gets created before Init() - it will have the wrong device count - so need someway to catch that -

Comment (GuanLuo) [reply src/device_memory_tracker.cc:147]: That only happen if you intentionally do so. Again, Init() is called on BackendInitiaze() which is the first place that the backend code will be executed. MemoryUsage is tracking usage of the model / instance which happens much later. I don't think it is common user error to initialize MemoryUsage object on BackendInitiaze, or actually not calling Init() as the first action when using the utility. I can see the user may forget to call `Init()` entirely, then the exception will be thrown anyway when they are calling the track function.

Comment (nnshah1) [reply src/device_memory_tracker.cc:147]: what happens if we have the throw? Trying to understand - this is so we can use memory usage without calling init()? (so calls can be left in?)

Comment (GuanLuo) [reply src/device_memory_tracker.cc:147]: Actually I think we can throw, I have changed the implementation locally to handle disabled case without too much conditionals and in such case the memory usage object won't be initialized.

File: include/triton/backend/device_memory_tracker.h [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+#include <vector>
+#include <algorithm>
+#include <functional>
+#include "triton/common/logging.h"
+#include "triton/core/tritonserver.h"
+#include "triton/backend/backend_common.h"
+
+#ifdef TRITON_ENABLE_GPU
+#include <cupti.h>
+#endif
```
Comment (rmccorm4) [inline include/triton/backend/device_memory_tracker.h:41]: Do we have a follow-up ticket to support this for CPU only server build? Seems like system/pinned mem usage would still be useful for cpu-only case, though understandably not as prioritized.

File: include/triton/backend/device_memory_tracker.h [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+/// only supports single client/subscriber. This is an known limitation and as a
+/// result, the memory tracker can cause unexpected application failure if other
+/// component of the Triton process also uses CUPTI with a different
+/// configuration, for example, the framework used by the backend may have
+/// implemented similar profiler with CUPTI. Therefore, before enabling this
+/// memory tracker utilities, you should make sure that there is no other CUPTI
+/// client in the process. This tracker is implemented with the assumption that
+/// all other possible CUPTI clients are using the same implementation so that
+/// as long as all backends are compiled with this memory tracker and they may
+/// interact with CUPTI initialized externally to the backend without issues.
```
Comment (rmccorm4) [inline include/triton/backend/device_memory_tracker.h:67]: This sentence reads weirdly to me. How does this sound?
```suggestion
/// client in the process. This tracker is implemented with the assumption that
/// all other CUPTI clients are using the same implementation so that
/// as long as all backends are compiled with this memory tracker, they may
/// interact with an externally-initialized CUPTI without issues.
```

File: include/triton/backend/device_memory_tracker.h [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+/// On TRITONBACKEND_Initialize
+///  - Call DeviceMemoryTracker::Init
+///
+/// If returns true, DeviceMemoryTracker::TrackThreadMemoryUsage and
+/// DeviceMemoryTracker::UntrackThreadMemoryUsage can be called accordingly to
+/// track memory allocation in the scope between the function, the memory usage
+/// will be recorded in MemoryUsage object and may be reported through
+/// TRITONBACKEND_ModelReportMemoryUsage or
+/// TRITONBACKEND_ModelInstanceReportMemoryUsage based on the entity of the
+/// memory usage.
```
Comment (rmccorm4) [inline include/triton/backend/device_memory_tracker.h:84]:
```suggestion
/// If Init returns true, DeviceMemoryTracker::TrackThreadMemoryUsage and
/// DeviceMemoryTracker::UntrackThreadMemoryUsage can be called accordingly to
/// track memory allocation in the scope between the two calls. The memory usage
/// will be recorded in MemoryUsage object and may be reported through
/// TRITONBACKEND_ModelReportMemoryUsage or
/// TRITONBACKEND_ModelInstanceReportMemoryUsage based on the entity of the
/// memory usage.
```

File: src/device_memory_tracker.cc [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+#include <stdexcept>
+#include "triton/core/tritonserver.h"
+
+namespace triton { namespace backend {
+
+std::unique_ptr<DeviceMemoryTracker> DeviceMemoryTracker::tracker_{nullptr};
+// Boilerplate from CUPTI examples
+namespace {
+
+// [WIP] figure out recover policy at different location of calling CUPTI
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:39]: Comment still needed? Follow-up ticket needed?

File: src/device_memory_tracker.cc [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+
+    // report any records dropped from the queue
+    size_t dropped;
+    CUPTI_CALL(cuptiActivityGetNumDroppedRecords(ctx, streamId, &dropped));
+    if (dropped != 0) {
+      LOG_WARNING << "Dropped " << dropped << " activity records";
+    }
+  }
+
+  free(buffer);
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:99]: protect with `if (buffer)` ?

File: src/device_memory_tracker.cc [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+    break;
+  }
+  case CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED: {
+    std::cerr << "CUPTI has been initialized elsewhere, assuming the implementation is the same" << std::endl;
+    break;
+  }
+  default:
+    // other error, should propagate and disable memory tracking for the backend
+    throw std::runtime_error("Unexpected failure on configuring CUPTI.");
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:136]: include return code `cupti_res` for easier debug

File: src/device_memory_tracker.cc [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+  }
+}
+
+int
+DeviceMemoryTracker::CudaDeviceCount()
+{
+  if (tracker_) {
+    return tracker_->device_cnt_;
+  }
+  throw std::runtime_error("DeviceMeoryTracker::Init() must be called before using any DeviceMeoryTracker features.");
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:146]:
```suggestion
  throw std::runtime_error("DeviceMeoryTracker::Init() must be called before using any DeviceMemoryTracker features.");
```

File: src/device_memory_tracker.cc [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+
+bool
+DeviceMemoryTracker::Init()
+{
+  if (tracker_ == nullptr) {
+    try {
+      tracker_.reset(new DeviceMemoryTracker());
+    } catch (const std::runtime_error& ex) {
+      // Fail initialization 
+      return false;
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:157]: LOG_ERROR `ex.what()` ?

File: src/device_memory_tracker.cc [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+void
+bufferRequested(uint8_t** buffer, size_t* size, size_t* maxNumRecords)
+{
+  uint8_t* bfr = (uint8_t*)malloc(BUF_SIZE + ALIGN_SIZE);
+  if (bfr != nullptr) {
+    *size = BUF_SIZE;
+    *buffer = ALIGN_BUFFER(bfr, ALIGN_SIZE);
+    *maxNumRecords = 0;
+  } else {
+    LOG_ERROR << "Failed to allocate buffer for CUPIT: out of memory";
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:67]:
```suggestion
    LOG_ERROR << "Failed to allocate buffer for CUPTI: out of memory";
```

File: src/device_memory_tracker.cc [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+  if (!usage) {
+    return;
+  }
+  if (tracker_) {
+    CUPTI_CALL(cuptiActivityPushExternalCorrelationId(
+        CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN,
+        reinterpret_cast<uint64_t>(&usage->cupti_tracker_)));
+    usage->tracked_ = true;
+  } else {
+    throw std::runtime_error("DeviceMeoryTracker::Init() must be called before using any DeviceMeoryTracker features.");
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:175]:
```suggestion
    throw std::runtime_error("DeviceMemoryTracker::Init() must be called before using any DeviceMemoryTracker features.");
```

File: src/device_memory_tracker.cc [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+  case CUPTI_SUCCESS : {
+    CUPTI_CALL(
+        cuptiActivityRegisterCallbacks(bufferRequested, bufferCompleted));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_MEMORY2));
+    CUPTI_CALL(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION));
+    break;
+  }
+  case CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED: {
+    std::cerr << "CUPTI has been initialized elsewhere, assuming the implementation is the same" << std::endl;
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:131]: `LOG_WARNING` instead of `std:cerr`?

File: include/triton/backend/device_memory_tracker.h (hunk 14 above)
Comment (GuanLuo) [reply include/triton/backend/device_memory_tracker.h:41]: I don't think we will plan that unless there is way to track it without injecting malloc

**Review by nnshah1 — APPROVED** [review]

File: src/device_memory_tracker.cc (hunk 18 above)
Comment (nnshah1) [reply src/device_memory_tracker.cc:109]: I believe free(0) is guranteed to be safe -

File: src/device_memory_tracker.cc [hunk 25]
```diff
@@ -120,20 +130,23 @@ DeviceMemoryTracker::DeviceMemoryTracker()
...
   case CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED: {
-    std::cerr << "CUPTI has been initialized elsewhere, assuming the implementation is the same" << std::endl;
+    LOG_WARNING << "CUPTI has been initialized elsewhere, assuming the implementation is the same";
     break;
   }
-  default:
+  default: {
     // other error, should propagate and disable memory tracking for the backend
-    throw std::runtime_error("Unexpected failure on configuring CUPTI.");
+    const char* errstr;\
```
Comment (rmccorm4) [inline src/device_memory_tracker.cc:146]:
```suggestion
    const char* errstr;
```

File: src/device_memory_tracker.cc (hunk 18 above)
Comment (rmccorm4) [reply src/device_memory_tracker.cc:109]: Today I learned!

File: CMakeLists.txt [hunk 26]
```diff
@@ -66,7 +66,11 @@ FetchContent_MakeAvailable(repo-common repo-core)
 #
 if(${TRITON_ENABLE_GPU})
   set(CUDA_USE_STATIC_CUDA_RUNTIME OFF)
+  set(CMAKE_FIND_DEBUG_MODE TRUE)
```
Comment (tanmayv25) [inline CMakeLists.txt:69]: Why was this needed here? Should we be using FIND_DEBUG_MODE everywhere we are using find_package as well?

File: include/triton/backend/device_memory_tracker.h [hunk 27]
```diff
@@ -0,0 +1,303 @@
...
+
+static_assert(
+    sizeof(uint64_t) >= sizeof(uintptr_t),
+    "The implementation is storing address pointer as uint64_t, "
+    "must ensure the space for pointer is <= sizeof(uint64_t).");
+
+namespace triton { namespace backend {
+
+/// DeviceMemoryTracker is a backend utility provided to track the memory
+/// allocated for a particular model and associated model instance. This utility
```
Comment (tanmayv25) [inline include/triton/backend/device_memory_tracker.h:51]: model instances

File: CMakeLists.txt (hunk 26 above)
Comment (GuanLuo) [reply CMakeLists.txt:69]: For debugging Windows build issue, not to be merged in.

**Review by tanmayv25 — APPROVED** [review]

**Review by Tabrizian — APPROVED** [review]

**Review by tanmayv25 — DISMISSED** [review]

**Review by Tabrizian — APPROVED** [review]

**Review by tanmayv25 — APPROVED** [review]
